saos-common/src/main/java/pl/edu/icm/saos/common/json/JsonUtils.java

```diff
@@ -5,49 +5,103 @@
      * 
-     * Returns next node as string. The next node is a node that is an array element or the whole object
-     * in case there is one element only (not array of elements).<br/>
-     * Returns null if there is no next node, for example when the traversing of array has reached an end.
-     * <br/>   
      * 
      * @throws IllegalArgumentException if the {@link JsonParser#getCodec()} == null. You can use {@link MappingJsonFactory} for creating
      * parser with default codec
-     * @throws JsonParseException if the nextNode cannot be parsed
+     * @throws JsonParseException if the nextToken cannot be parsed
      * @throws IOException if I/O error occured
      * 
      * */
-    public static String nextNode(JsonParser jp) throws IOException {
+    public JsonToken nextJsonObject(JsonParser jsonParser) throws JsonParseException, IOException {
+        
+        Preconditions.checkArgument(jsonParser.getCodec() != null);
+        
+        JsonToken token = jsonParser.nextToken();
+        
+        if (isJsonObject(token)) {
         
-        Preconditions.checkArgument(jp.getCodec() != null);
+            return token;
         
-        JsonToken current = jp.nextToken();
-        if (current == JsonToken.START_ARRAY) {
-           jp.nextToken();
-        } else if (current == JsonToken.END_ARRAY) {
+        } else {
+            
             return null;
+            
         }
+            
+    }
+    
+    /**
+     * Returns the next json object as object of the given objectClass.
+     * <ul>Invokes:
+     * <li>{@link #nextJsonObject(JsonParser)}</li>
+     * <li>{@link JsonParser#readValueAs(Class)}</li>
+     * </ul>
+     * Returns null if {@link #nextJsonObject(JsonParser)} returns null.
+     *
+     */
+    public <T> T nextJsonObject(JsonParser jsonParser, Class<T> objectClass) throws IOException {
+        
+        JsonToken token = nextJsonObject(jsonParser);
         
-        JsonNode jsonNode = jp.readValueAsTree();
-        return jsonNode == null? null: jsonNode.toString();
+        if (token == null) {
             
+            return null;
+        
+        }
+        
+        return jsonParser.readValueAs(objectClass);
+    
     }
     
+    
+    /**
+     * Formats the current token. Invokes {@link JsonParser#readValueAsTree()}<br/>
+     * Returns null if {@link JsonParser#readValueAsTree()} returns {@link NullNode}
+     * @throws JsonParseException if the current token is not a valid json
+     * @throws IOException in case of I/O Error
+     */
+    public String formatCurrentTokenTree(JsonParser jsonParser) throws IOException {
+
+        TreeNode node = jsonParser.readValueAsTree();
+        
+        if (node instanceof NullNode) {
+            return null;
+        }
+        
+        return node.toString();
+    
+    } 
+    
+    
+    /**
+     * Does the given jsonToken indicate the start of json object or array?
+     */
+    public boolean isJsonObject(JsonToken jsonToken) {
```

Comment:
I'll leave it. I think it fits nicely to this class, especially to nextJsonObject methods.